### PR TITLE
Improve Error Handling

### PR DESCRIPTION
**Pull request by kkoutsilis**
In this PR we improve the error handling mechanism by using [RunE](https://pkg.go.dev/github.com/spf13/cobra#Command.RunE) that returns an error instead of [Run](https://pkg.go.dev/github.com/spf13/cobra#Command.Run) , and delegating the error handling to the caller instead of `log.Fatal()` every time there is an error in the code.

### CODE REVIEW

**Comment by manavo** [comment]
This should return the error instead of calling `log.Fatalf`: https://github.com/kkoutsilis/SSMG/blob/f1f066a1c99a591d0bff6cc707dc23699b29eb6f/cmd/root.go#L94C13-L94C13 (the `sendEmails` signature would need to change to return an error)

And then handle the error in the same way here: https://github.com/kkoutsilis/SSMG/pull/8/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fL152

**Comment by kkoutsilis** [comment]
> This should return the error instead of calling `log.Fatalf`: https://github.com/kkoutsilis/SSMG/blob/f1f066a1c99a591d0bff6cc707dc23699b29eb6f/cmd/root.go#L94C13-L94C13 (the `sendEmails` signature would need to change to return an error)
> 
> And then handle the error in the same way here: https://github.com/kkoutsilis/SSMG/pull/8/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fL152

Have a look now :)

**Comment by manavo** [comment]
Looking better!

The only other consideration is that if one of the emails fails, the process exits rather than continuing.

Might be worth saving how many fail. If they all succeed, there's no error to return. If at least one succeeds, return an error which shows how many (and which ones?) failed?

**Comment by kkoutsilis** [comment]
> Looking better!
> 
> The only other consideration is that if one of the emails fails, the process exits rather than continuing.
> 
> Might be worth saving how many fail. If they all succeed, there's no error to return. If at least one succeeds, return an error which shows how many (and which ones?) failed?

Yeah, this needs to be improved

**Comment by kkoutsilis** [comment]
> Looking better!
> 
> The only other consideration is that if one of the emails fails, the process exits rather than continuing.
> 
> Might be worth saving how many fail. If they all succeed, there's no error to return. If at least one succeeds, return an error which shows how many (and which ones?) failed?

@manavo Added a temp solution that does not stop the execution if there is an error sending some email, will work on a better solution later on.

File: cmd/root.go [hunk 1]
```diff
@@ -66,13 +66,13 @@ func sendEmail(to, subject, body string) error {
...
 	if err != nil {
-		return err
+		return fmt.Errorf("error reading email server port %w", err)
 	}
 
 	d := gomail.NewDialer(host, port, user, password)
 	s, err := d.Dial()
 	if err != nil {
-		return err
+		return fmt.Errorf("error dialing email server %w", err)
```
Comment (manavo) [inline cmd/root.go:75]: `errors.Wrap` might be good here (https://pkg.go.dev/github.com/pkg/errors#readme-adding-context-to-an-error)

File: cmd/root.go [hunk 2]
```diff
@@ -66,13 +66,13 @@ func sendEmail(to, subject, body string) error {
 
 	port, err := strconv.Atoi(strPort)
 	if err != nil {
-		return err
+		return fmt.Errorf("error reading email server port %w", err)
```
Comment (manavo) [inline cmd/root.go:69]: `errors.Wrap` here as well?

File: cmd/root.go [hunk 3]
```diff
@@ -84,36 +84,42 @@ func sendEmail(to, subject, body string) error {
 	return gomail.Send(s, m)
 }
 
-func sendEmails(matches []MatchPair) {
+func sendEmails(matches []MatchPair) ([]string, error) {
+	// TODO: add retry logic for failed emails (maybe use a queue)
+	// instead of sending each email individually,
+	// generate all the emails and send them in bulk to avoid individual errors
```
Comment (manavo) [inline cmd/root.go:90]: The simpler option here might be to prompt the user if they want to retry?

File: cmd/root.go [hunk 4]
```diff
@@ -84,36 +84,42 @@ func sendEmail(to, subject, body string) error {
...
+	emailIssues := []string{}
 	subject := "Your Secret Santa Match!"
 
 	templateFile := "./templates/email_template.html"
 	tmpl, err := template.ParseFiles(templateFile)
 	if err != nil {
-		log.Fatalf("Error parsing email template: %v", err)
+		return emailIssues, fmt.Errorf("failed to parse email template: %w", err)
```
Comment (manavo) [inline cmd/root.go:97]: `errors.Wrap` again?

File: cmd/root.go [hunk 5]
```diff
@@ -84,36 +84,42 @@ func sendEmail(to, subject, body string) error {
 	return gomail.Send(s, m)
 }
 
-func sendEmails(matches []MatchPair) {
+func sendEmails(matches []MatchPair) ([]string, error) {
+	// TODO: add retry logic for failed emails (maybe use a queue)
+	// instead of sending each email individually,
+	// generate all the emails and send them in bulk to avoid individual errors
+	emailIssues := []string{}
```
Comment (manavo) [inline cmd/root.go:91]: You could pre-allocate the capacity of this. Insignificant micro-optimization in this case, but good habit.

```go
emailIssues := make([]string, 0, len(matches))
```

I think the above is correct, but untested

File: cmd/root.go [hunk 6]
```diff
@@ -84,36 +84,42 @@ func sendEmail(to, subject, body string) error {
...
+		return emailIssues, fmt.Errorf("failed to parse email template: %w", err)
 	}
 
 	for _, match := range matches {
 		var emailBodyBuffer = &strings.Builder{}
 		err := tmpl.Execute(emailBodyBuffer, match)
 		if err != nil {
-			log.Printf("Error executing template for %s: %v", match.From.Email, err)
-			continue
+			emailIssues = append(emailIssues, fmt.Sprintf("failed to execute template for %s: %v", match.From.Email, err))
```
Comment (manavo) [inline cmd/root.go:104]: If we do have an issue here, we actually still want to call `continue`, no point running the rest of the code in this iteration?

File: cmd/root.go [hunk 7]
```diff
@@ -125,31 +131,41 @@ var rootCmd = &cobra.Command{
 		}
 
 		if !checkFileExists(filePath) {
-			log.Fatal("File ", filePath, " does not exist")
+			return fmt.Errorf("file %s does not exist", filePath)
```
Comment (manavo) [inline cmd/root.go:134]: `errors.Wrap` here and on all the ones below?

Comment (kkoutsilis) [reply cmd/root.go:134]: This one does not have an err to wrap, but the rest of them are updated

**Review by manavo — APPROVED** [review]

**Review by manavo — APPROVED** [review]

**Pull request merged**